twenty-first/src/util_types/algebraic_prng.rs

```diff
@@ -0,0 +1,37 @@
+use itertools::Itertools;
+
+use crate::shared_math::x_field_element::XFieldElement;
+
+use super::algebraic_hasher::SpongeHasher;
+
+pub trait AlgebraicPrng {
+    type PrngState;
+    fn sample_xfield(state: &mut Self::PrngState, num_elements: usize) -> Vec<XFieldElement>;
+    fn sample_indices(state: &mut Self::PrngState, max: u32, num_elements: usize) -> Vec<u32>;
+}
+
+impl<H: SpongeHasher> AlgebraicPrng for H {
+    type PrngState = H::SpongeState;
+
+    fn sample_xfield(state: &mut Self::PrngState, num_elements: usize) -> Vec<XFieldElement> {
+        let num_squeezes =
+            (num_elements * 3 + <H as SpongeHasher>::RATE - 1) / <H as SpongeHasher>::RATE;
+        (0..num_squeezes)
+            .into_iter()
+            .flat_map(|_| <H as SpongeHasher>::squeeze(state))
+            .collect_vec()
+            .chunks(3)
+            .map(|ch| XFieldElement::new(ch.try_into().unwrap()))
+            .collect_vec()
+    }
```

Comment:
The reason `sample_indices()` and `sample_weights()` were recently removed from twenty-first is because the way weights are sampled is highly specific to Triton VM; that is, whatever way is most convenient to sample weights inside the VM should affect how we sample them in STARK/FRI (which are also in Triton VM).

I had originally placed them on this trait because I wanted to bundle everything related to hashing in one place. Ferdinand convinced me that it's probably better to have them (or at least `sample_weights()`) in Triton VM. (You could similarly argue that `squeeze` and `absorb` are quite coupled to Triton VM as well.)

If I understand correctly, we have two independent subscribers to `sample_indices()`, but only one subscriber to `sample_weights()`, which is Triton VM. If that is true, we might consider only having `sample_indices()` in twenty-first.